sound/soc/sof/ipc4-pcm.c

```diff
@@ -495,17 +496,44 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
 				use_chain_dma = true;
 		}
 	}
+
+	available_fmt = &ipc4_copier->available_fmt;
+	fe_rate = params_rate(params);
+
 	/*
-	 * Always set BE format to 32-bits for both playback and capture, except if chained
-	 * DMA is in use. In chained DMA case the back-end and front-end format must match.
+	 * Limit sampling rate if copier has a format that matches
+	 * the FE rate.
 	 */
 	if (!use_chain_dma) {
-		snd_mask_none(fmt);
-		snd_mask_set_format(fmt, SNDRV_PCM_FORMAT_S32_LE);
-	}
+		struct sof_ipc4_audio_format *audio_fmt =
+			&available_fmt->input_pin_fmts[0].audio_fmt;
+		unsigned int be_rate  = audio_fmt->sampling_frequency;
+		bool fe_be_rate_match = false;
+		bool single_be_rate = true;
+		int i;
+
+		/*
+		 * copier does not change sampling rate, so we
+		 * need to only check input formats
+		 */
+		for (i = 0; i < available_fmt->num_input_formats; i++) {
+			audio_fmt = &available_fmt->input_pin_fmts[i].audio_fmt;
 
-	rate->min = ipc4_copier->available_fmt.input_pin_fmts->audio_fmt.sampling_frequency;
-	rate->max = rate->min;
+			if (audio_fmt->sampling_frequency == fe_rate) {
+				fe_be_rate_match = true;
+				break;
+			}
+
+			if (audio_fmt->sampling_frequency != be_rate)
+				single_be_rate = false;
+		}
+
+		if (!fe_be_rate_match && single_be_rate) {
+			/* Only one format defined, use that */
+			rate->min = available_fmt->input_pin_fmts[0].audio_fmt.sampling_frequency;
+			rate->max = rate->min;
```

Comment:
It took me some time to get this, but the comment does not helped...
And I'm still not sure if I got it...

`available_fmt->input_pin_fmts[0].audio_fmt.sampling_frequency` is `be_rate`, right?

no, I don't get what this is doing.

What happens if the available_fmt->input_pin_fmts[] only contains single rates, but different sample sizes, channels and this rate does not match for the fe_rate?
We will fail somewhere else?

Reply:
@ujfalusi Right, now that I moved (in V11) this piece of code to same scope where 'be_rate'  is defined, yes, I could assign to 'be_rate' now. Will fix the comment.

For the bigger question of what happens if rate is not limited at all, that's a good question. I've been just going by what the existing code does and what we've done with ipc3-pcm.c in the past. For the specific need of enabling multiple BE configs, I've not wanted to reopen the wider arch.

But there is certainly overlap. The dai-fixup is called in hw_params flow. This is standard ASoC stuff. Then in sof_pcm_hw_params() we run the sof_pcm_setup_connected_widgets() path which will do a more complex path enablement and this logic is more specific to SOF. E.g. for capture we traverse all widgets on the path, and set formats for each.

Now whether all constraints could be moved to sof_pcm_setup_connected_widgets(), I don't really know. I don't really see a big obstacle really. @ranj063 @plbossart ? Both paths of logic are done at hw_params epoch, so I don't see a big difference, but I'm sure there is some subtle dependencies.

Reply:
@ujfalusi Dug the history of this code a bit more and the key reason to have the "rate->min/max" lookup in dai fixup is that we do the SSP blob selection here as well (via ipc4_ssp_dai_config_pcm_params_match()). So that's one option to hide this logic behind the SSP specific code and only limit the rate if we need to choose a blob. At least this would be hidden deeper in vendor/DAI specific code and not have this in generic fixup code for IPC4....

Reply:
@kv2019i Another reason is for SRC, .e.g. 44.1k stream for FE, we do SRC in fw to convert it to 48k supported by BE, so we need to use this fixup to change rate for audio bus controller and codec driver.

Reply:
Maybe it is caused my first patch, now I prefer to use output format available_fmt->output_pin_fmts[0].audio_fmt.sampling_frequency since output is for audio bus and codec. we may have different input valid bit from internal pipeline but only one output valid bits by audio dsp

Reply:
@RanderWang fixup is called for both directions (capture/playback) so use of copier pins configs is problematic as e.g. for playback case, copier input may be different than what is actually sent to the DAI. Especially to evaluate sample formats (as copier can do conversions and input/output formats may be different), this is a real problem. For sampling rate (as used in this PR) and channel count, it should not matter whether we look at input or output pins, as both must be the same (as copier cannot convert). This is harder to follow in general than IPC3, where we had an explicit object to describe DAI properties.